pkg/services/serviceaccounts/proxy/service.go

```diff
@@ -0,0 +1,100 @@
+package proxy
+
+import (
+	"context"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/apikey"
+	"github.com/grafana/grafana/pkg/services/extsvcauth/extsvcaccounts"
+	"github.com/grafana/grafana/pkg/services/serviceaccounts"
+	"github.com/grafana/grafana/pkg/services/serviceaccounts/manager"
+)
+
+// ServiceAccountsProxy is a proxy for the serviceaccounts.Service interface
+// that is used to add validations to service accounts and protects external
+// service accounts from being modified by users.
+
+type ServiceAccountsProxy struct {
+	log            log.Logger
+	proxiedService serviceaccounts.Service
+}
+
+func ProvideServiceAccountsProxy(
+	proxiedService *manager.ServiceAccountsService,
+) (*ServiceAccountsProxy, error) {
+	s := &ServiceAccountsProxy{
+		log:            log.New("serviceaccounts.proxy"),
+		proxiedService: proxiedService,
+	}
+	return s, nil
+}
+
+var _ serviceaccounts.Service = (*ServiceAccountsProxy)(nil)
+
+func (s *ServiceAccountsProxy) CreateServiceAccount(ctx context.Context, orgID int64, saForm *serviceaccounts.CreateServiceAccountForm) (*serviceaccounts.ServiceAccountDTO, error) {
+	if isExternalServiceAccount(saForm.Name) {
+		return nil, extsvcaccounts.ErrExtServiceAccountInvalidName
+	}
+	return s.proxiedService.CreateServiceAccount(ctx, orgID, saForm)
+}
+
+func (s *ServiceAccountsProxy) DeleteServiceAccount(ctx context.Context, orgID, serviceAccountID int64) error {
+	sa, err := s.proxiedService.RetrieveServiceAccount(ctx, orgID, serviceAccountID)
+	if err != nil {
+		return err
+	}
+
+	if isExternalServiceAccount(sa.Login) {
+		return extsvcaccounts.ErrExtServiceAccountCannotBeDeleted
+	}
+
+	return s.proxiedService.DeleteServiceAccount(ctx, orgID, serviceAccountID)
+}
+
+func (s *ServiceAccountsProxy) RetrieveServiceAccount(ctx context.Context, orgID, serviceAccountID int64) (*serviceaccounts.ServiceAccountProfileDTO, error) {
+	sa, err := s.proxiedService.RetrieveServiceAccount(ctx, orgID, serviceAccountID)
+	if err != nil {
+		return nil, err
+	}
+
+	sa.IsExternal = isExternalServiceAccount(sa.Login)
+
+	return sa, nil
+}
+
+func (s *ServiceAccountsProxy) RetrieveServiceAccountIdByName(ctx context.Context, orgID int64, name string) (int64, error) {
+	return s.proxiedService.RetrieveServiceAccountIdByName(ctx, orgID, name)
```

Comment:
Should we set `IsExternal` here as well?

Reply:
Not really, it will be ignored  since the return value is `int64`

Reply:
My bad, I didn't check the return type :+1: